### PR TITLE
Several perf improvements around shape processing

### PR DESCRIPTION
**Pull request by sebastienros**

### CODE REVIEW

File: src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs [hunk 1]
```diff
@@ -27,23 +30,28 @@ public class TemplatesShapeBindingResolver : IShapeBindingResolver
         {
             _templatesManager = templatesManager;
             _liquidTemplateManager = liquidTemplateManager;
-            _previewTemplatesProvider = previewTemplatesProvider;
             _httpContextAccessor = httpContextAccessor;
             _htmlEncoder = htmlEncoder;
+            _localTemplates = previewTemplatesProvider.GetTemplates();
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs:35]: Should we add `Lazy` here?

File: src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs [hunk 2]
```diff
@@ -27,23 +30,28 @@ public class TemplatesShapeBindingResolver : IShapeBindingResolver
...
+
+            if (isAdmin.Value)
             {
                 return null;
             }
 
-            var localTemplates = _previewTemplatesProvider.GetTemplates();
-
-            if (localTemplates != null && localTemplates.Templates.TryGetValue(shapeType, out var localTemplate))
+            if (_localTemplates != null && _localTemplates.Templates.Any())
```
Comment (MikeAlhayek) [inline src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs:49]: simpler with this

```
            if (_localTemplates != null && 
            _localTemplates.Templates.Count > 0  &&
             _localTemplates.Templates.TryGetValue(shapeType, out var localTemplate))
            {
                return BuildShapeBinding(shapeType, localTemplate);
            }
```

File: src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs [hunk 3]
```diff
@@ -105,10 +105,8 @@ public static TagBuilder GetTagBuilder(this IShape shape, string defaultTagName
 
             if (shape.Classes != null)
             {
-                foreach (var cssClass in shape.Classes)
-                {
-                    tagBuilder.AddCssClass(cssClass);
-                }
+                // Faster than AddCssClass which will do twice as many concatenations as classes.
+                tagBuilder.Attributes["class"] = string.Join(' ', shape.Classes);
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs:109]: maybe event `shape.Classes != null && shape.Classes.Length > 0` so we don't call it at all if we don't have anything

File: src/OrchardCore/OrchardCore.DisplayManagement/Implementation/DefaultHtmlDisplay.cs [hunk 4]
```diff
@@ -237,44 +239,50 @@ private async Task<ShapeBinding> GetShapeBindingAsync(string shapeType, Alternat
...
             // When no alternates matches, the shapeType is used to find the longest matching binding,
-            // the shape-type name can break itself into shorter fallbacks at double-underscore marks,
-            // so the shape-type itself may contain a longer alternate forms that falls back to a shorter one.
-            var shapeTypeScan = shapeType;
+            // the shapetype name can break itself into shorter fallbacks at double-underscore marks,
+            // so the shapetype itself may contain a longer alternate forms that falls back to a shorter one.
 
-            do
+            // Build a cache of such values
+            var alternateShapeTypes = _alternateShapeTypes.GetOrAdd(shapeType, shapeType =>
```
Comment (MikeAlhayek) [inline src/OrchardCore/OrchardCore.DisplayManagement/Implementation/DefaultHtmlDisplay.cs:246]: Did you benchmark this change?

**Review by MikeAlhayek — CHANGES_REQUESTED** [review]

File: src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs (hunk 2 above)
Comment (MikeAlhayek) [reply src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs:49]: I don't even think `_localTemplates.Templates.Count > 0` is necessary

File: src/OrchardCore/OrchardCore.DisplayManagement/Implementation/DefaultHtmlDisplay.cs (hunk 4 above)
Comment (sebastienros) [reply src/OrchardCore/OrchardCore.DisplayManagement/Implementation/DefaultHtmlDisplay.cs:246]: With bombardier locally by removing other components, and it showed. Also I did it as it appeared as a main factor of string concatenation, then looked into the TagHelper implementation which is creating new string of every call.

File: src/OrchardCore.Modules/OrchardCore.Templates/Services/AdminTemplatesShapeBindingResolver.cs [hunk 5]
```diff
@@ -17,6 +17,7 @@ public class AdminTemplatesShapeBindingResolver : IShapeBindingResolver
         private readonly AdminPreviewTemplatesProvider _previewTemplatesProvider;
         private readonly IHttpContextAccessor _httpContextAccessor;
         private readonly HtmlEncoder _htmlEncoder;
+        private bool? isAdmin;
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Templates/Services/AdminTemplatesShapeBindingResolver.cs:20]:
```suggestion

        private bool? _isAdmin;
```

File: src/OrchardCore.Modules/OrchardCore.Templates/Services/AdminTemplatesShapeBindingResolver.cs [hunk 6]
```diff
@@ -34,7 +35,11 @@ public class AdminTemplatesShapeBindingResolver : IShapeBindingResolver
 
         public async Task<ShapeBinding> GetShapeBindingAsync(string shapeType)
         {
-            if (!AdminAttribute.IsApplied(_httpContextAccessor.HttpContext))
+            // Cache this value since the service is scoped and this method is invoked for every
+            // alternate of every shape.
+            isAdmin ??= AdminAttribute.IsApplied(_httpContextAccessor.HttpContext);
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Templates/Services/AdminTemplatesShapeBindingResolver.cs:40]:
```suggestion
            _isAdmin ??= AdminAttribute.IsApplied(_httpContextAccessor.HttpContext);
```

File: src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs [hunk 7]
```diff
@@ -12,11 +13,13 @@ namespace OrchardCore.Templates.Services
...
     {
         private TemplatesDocument _templatesDocument;
+        private readonly TemplatesDocument _localTemplates;
+
         private readonly TemplatesManager _templatesManager;
         private readonly ILiquidTemplateManager _liquidTemplateManager;
-        private readonly PreviewTemplatesProvider _previewTemplatesProvider;
         private readonly IHttpContextAccessor _httpContextAccessor;
         private readonly HtmlEncoder _htmlEncoder;
+        private bool? isAdmin;
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs:22]:
```suggestion

        private TemplatesDocument _templatesDocument;
        private bool? _isAdmin;
```

File: src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs [hunk 8]
```diff
@@ -12,11 +13,13 @@ namespace OrchardCore.Templates.Services
     public class TemplatesShapeBindingResolver : IShapeBindingResolver
     {
         private TemplatesDocument _templatesDocument;
+        private readonly TemplatesDocument _localTemplates;
+
```
Comment (hishamco) [inline src/OrchardCore.Modules/OrchardCore.Templates/Services/TemplatesShapeBindingResolver.cs:17]:
```suggestion
        private readonly TemplatesDocument _localTemplates;
```

**Review by hishamco — CHANGES_REQUESTED** [review]

File: src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs [hunk 9]
```diff
@@ -98,12 +99,12 @@ public static TagBuilder GetTagBuilder(this IShape shape, string defaultTagName
 
             var tagBuilder = new TagBuilder(tagName);
 
-            if (shape.Attributes != null)
+            if (shape.Attributes != null && shape.Attributes.Any())
```
Comment (lahma) [inline src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs:102]: `shape.Attributes.Count > 0` too fast so need to slow down with LINQ? 😢

Comment (sebastienros) [reply src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs:102]: @lahma you need to refresh your skills ;) It's generating the same IL these days. So LINQ all the way as it's easier to read IMO.

Comment (lahma) [reply src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs:102]: I need to get the compiler/JIT that you are using 😝 

https://sharplab.io/#v2:EYLgtghglgdgNAFxBAzmAPgAQAwAJMCMALANwCwAUDvgQHQAysAjuRZZgMz4BMuAwrgDelXKNwixASQAiUAMYIoAexgQATgE8APIWxwa2AHy4AgggRqowAK4IApiiG4A5nYQlcAXwmjKP/FzASkoANrgAsgQAFACUuAC8xmYWVrYOuACE8bgw1iFhAGQFpuaWNvYotCYwGrGs/py4QaER3LEJSaWpFZnZufm4RSUp5Q60fErWMAi4xtj1FJ5AA==

Comment (sebastienros) [reply src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs:102]: I opened sharplab too when I realized these were on interfaces. You win!

Comment (lahma) [reply src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs:102]: 🥳

Comment (MikeAlhayek) [reply src/OrchardCore/OrchardCore.DisplayManagement/IShape.cs]: @sebastienros even if it is the same thing, I thing `.Count > 0` is nicer and should be used when available.

**Pull request merged**